fe/fe-core/src/main/java/com/starrocks/scheduler/mv/MVPCTRefreshRangePartitioner.java

```diff
@@ -0,0 +1,446 @@
+                mvContext.getMvRefBaseTableIntersectedPartitions();
+        for (String mvPartitionName : mvPartitionNames) {
+            if (mvRefBaseTablePartitionMaps == null || !mvRefBaseTablePartitionMaps.containsKey(mvPartitionName)) {
+                LOG.warn("Cannot find need refreshed mv table partition from synced partition info: {}",
+                        mvPartitionName);
+                continue;
+            }
+            Map<Table, Set<String>> mvRefBaseTablePartitionMap = mvRefBaseTablePartitionMaps.get(mvPartitionName);
+            for (Map.Entry<Table, Set<String>> entry : mvRefBaseTablePartitionMap.entrySet()) {
+                Table baseTable = entry.getKey();
+                Set<String> baseTablePartitions = entry.getValue();
+                // If the result already contains the base table name, add all new partitions to the existing set
+                // If the result doesn't contain the base table name, put the new set into the map
+                result.computeIfAbsent(baseTable, k -> Sets.newHashSet()).addAll(baseTablePartitions);
+            }
+        }
+        return result;
+    }
+
+    private void addRangePartitions(Database database, MaterializedView mv,
+                                    Map<String, Range<PartitionKey>> adds, Map<String, String> partitionProperties,
+                                    DistributionDesc distributionDesc) {
+        if (adds.isEmpty()) {
+            return;
+        }
+        List<PartitionDesc> partitionDescs = Lists.newArrayList();
+
+        for (Map.Entry<String, Range<PartitionKey>> addEntry : adds.entrySet()) {
+            String mvPartitionName = addEntry.getKey();
+            Range<PartitionKey> partitionKeyRange = addEntry.getValue();
+
+            String lowerBound = partitionKeyRange.lowerEndpoint().getKeys().get(0).getStringValue();
+            String upperBound = partitionKeyRange.upperEndpoint().getKeys().get(0).getStringValue();
+            boolean isMaxValue = partitionKeyRange.upperEndpoint().isMaxValue();
+            PartitionValue upperPartitionValue;
+            if (isMaxValue) {
+                upperPartitionValue = PartitionValue.MAX_VALUE;
+            } else {
+                upperPartitionValue = new PartitionValue(upperBound);
+            }
+            PartitionKeyDesc partitionKeyDesc = new PartitionKeyDesc(
+                    Collections.singletonList(new PartitionValue(lowerBound)),
+                    Collections.singletonList(upperPartitionValue));
+            SingleRangePartitionDesc singleRangePartitionDesc =
+                    new SingleRangePartitionDesc(false, mvPartitionName, partitionKeyDesc, partitionProperties);
+            partitionDescs.add(singleRangePartitionDesc);
+        }
+
+        // create partitions in small batch, to avoid create too many partitions at once
+        for (List<PartitionDesc> batch : ListUtils.partition(partitionDescs, CREATE_PARTITION_BATCH_SIZE)) {
```

Comment:
can we just cook a textual sql "ALTER TABLE {tableName} ADD PARTITION {parititon_definition}" to add partitions?

Reply:
I think it's better to use api rather than `sql` since it may add more extra overheads by using sql.